File: backend/src/main/kotlin/com/example/backend/controllers/DepenseController.kt
package com.example.backend.controllers

import com.example.backend.models.Depense
import com.example.backend.services.DepenseService
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/depenses")
class DepenseController(private val depenseService: DepenseService) {

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
    @GetMapping
    fun getAllDepenses(): ResponseEntity<List<Depense>> {
        return ResponseEntity.ok(depenseService.getAllDepenses())
    }
  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
    @GetMapping("/pageable")
    fun getAllDepensesPageable(
    @RequestParam(defaultValue = "0") page: String,
    @RequestParam(defaultValue = "10") size: String,
    @RequestParam(defaultValue = "id") sortBy: String,
    ): ResponseEntity<Page<Map<String, Any?>>> {
    val pageNumber = page.toIntOrNull()?.coerceAtLeast(0) ?: 0
    val pageSize = size.toIntOrNull()?.coerceAtLeast(1) ?: 10
    val pageable = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.DESC, "dateDepense"))
    return ResponseEntity.ok(depenseService.getAllDepensesPageabel(pageable))
    }

@PostMapping
fun createDepense(@RequestBody depenseData: Map<String, Any>): ResponseEntity<Depense> {
    val designation = depenseData["designation"] as? String
        ?: throw IllegalArgumentException("Missing or invalid 'designation'")
    val prixUnitaire = depenseData["prixUnitaire"] as? Int
        ?: throw IllegalArgumentException("Missing or invalid 'prixUnitaire'")
  if (depenseData.size>3){
    val depense = depenseService.createDepenseMap(depenseData)
    return ResponseEntity.ok(depense)
  }
  else {
    val depense = depenseService.createDepense(designation, prixUnitaire)
    return ResponseEntity.ok(depense)
  }

}
    @PutMapping("/{id}")
    fun updateDepense(@PathVariable id: Int,@RequestBody depenseData: Map<String, Any>): ResponseEntity<Depense> {
    val designation = depenseData["designation"] as? String
      ?: throw IllegalArgumentException("Missing or invalid 'designation'")
    val prixUnitaire = depenseData["prixUnitaire"] as? Int
      ?: throw IllegalArgumentException("Missing or invalid 'prixUnitaire'")
    val depense = depenseService.updateDepense(id,designation, prixUnitaire)
    return ResponseEntity.ok(depense)
    }

    @DeleteMapping("/{id}")
    fun deleteDepense(@PathVariable id: Int): ResponseEntity<Void> {
        depenseService.deleteDepense(id)
        return ResponseEntity.noContent().build()
    }
}

File: backend_py/pharmaPython/app/api/v1/depense_controller.py
# depense_controller.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Any, Dict, List

from app.api.deps import get_db
from app.models.depense import Depense, DepenseIn
from app.services.depense_service import DepenseService
from app.utility.jwt_authentication import jwt_authentication

router = APIRouter(
  prefix="/depenses",
  tags=["Depenses"],
  dependencies=[Depends(jwt_authentication)]  # équivalent de @PreAuthorize("isAuthenticated()")
)


@router.get("/")
def list_depenses(db: Session = Depends(get_db)):
  return DepenseService(db).get_all_depenses()


@router.get("/pageable")
def list_depenses_pageable(
  page: int = Query(0, ge=0),
  size: int = Query(10, ge=1),
  sortBy: str = Query("dateDepense"),
  db: Session = Depends(get_db),
):
  """
  Équivalent de: Page<Map<String, Any?>> avec tri DESC sur dateDepense par défaut.
  On renvoie un objet de pagination simple et la liste mappée.
  """
  full = DepenseService(db).get_all_depenses_pageable(page=page, size=size, sort=sortBy, direction="desc")

  return full


@router.post("/", response_model=DepenseIn)
def create(depenseData: Dict[str, Any], db: Session = Depends(get_db)):
  """
  Kotlin: si la map contient > 3 champs → createDepenseMap, sinon createDepense(designation, prixUnitaire)
  """
  designation = depenseData.get("designation")
  prix_unitaire = depenseData.get("prixUnitaire")

  if designation is None or not isinstance(designation, str):
    raise HTTPException(status_code=422, detail="Missing or invalid 'designation'")
  if prix_unitaire is None or not isinstance(prix_unitaire, int):
    raise HTTPException(status_code=422, detail="Missing or invalid 'prixUnitaire'")

  if len(depenseData) > 3:
    return DepenseService(db).create_depense_map(db, depenseData)
  else:
    return DepenseService(db).create_depense(designation, prix_unitaire)


@router.put("/{id}", response_model=DepenseIn)
def update(id: int, depenseData: Dict[str, Any], db: Session = Depends(get_db)):
  designation = depenseData.get("designation")
  prix_unitaire = depenseData.get("prixUnitaire")

  if designation is None or not isinstance(designation, str):
    raise HTTPException(status_code=422, detail="Missing or invalid 'designation'")
  if prix_unitaire is None or not isinstance(prix_unitaire, int):
    raise HTTPException(status_code=422, detail="Missing or invalid 'prixUnitaire'")

  return DepenseService(db).update_depense(id, designation, prix_unitaire)


@router.delete("/{id}", status_code=204)
def delete(id: int, db: Session = Depends(get_db)):
  # tu peux faire une suppression logique si ton modèle la supporte
  dep:Depense = db.query(Depense).get(id)
  dep.supprimer = 1
  if not dep:
    raise HTTPException(status_code=404, detail="Dépense introuvable")
  db.commit()
  return {"message": "Deleted"}
